x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheDirectory.java

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.searchablesnapshots.cache;
+
+import org.apache.lucene.store.AlreadyClosedException;
+import org.apache.lucene.store.BufferedIndexInput;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.FilterDirectory;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.SuppressForbidden;
+import org.elasticsearch.common.io.Channels;
+import org.elasticsearch.common.util.concurrent.ReleasableLock;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * {@link CacheDirectory} uses a {@link CacheService} to cache Lucene files provided by another {@link Directory}.
+ */
+public class CacheDirectory extends FilterDirectory {
+
+    private final CacheService cacheService;
+    private final Path cacheDir;
+
+    public CacheDirectory(Directory in, CacheService cacheService, Path cacheDir) throws IOException {
+        super(in);
+        this.cacheService = Objects.requireNonNull(cacheService);
+        this.cacheDir = Files.createDirectories(cacheDir);
+    }
+
+    public void close() throws IOException {
+        super.close();
+        // Ideally we could let the cache evict/remove cached files by itself after the
+        // directory has been closed.
+        cacheService.removeFromCache(key -> key.startsWith(cacheDir.toString()));
+    }
+
+    @Override
+    public IndexInput openInput(final String name, final IOContext context) throws IOException {
+        ensureOpen();
+        return new CacheBufferedIndexInput(name, fileLength(name), context);
+    }
+
+    public class CacheBufferedIndexInput extends BufferedIndexInput implements CacheFile.EvictionListener {
+
+        private final String fileName;
+        private final long fileLength;
+        private final Path file;
+        private final IOContext ioContext;
+        private final long offset;
+        private final long end;
+
+        private @Nullable AtomicReference<CacheFile> cacheFile;
+        private @Nullable CacheBufferedIndexInput parent;
+        private ReleasableLock cacheEvictionLock;
```

Comment:
I think we don't need a special kind of lock here any more, `synchronized (this)` should be ok here?

Reply:
Right, `synchronized` blocks should be enough. I pushed 7227422525f67a8c4f55e8511220404db374662d